File: tests/fixtures/selenium-chaos-suite/src/core/DriverFactory.ts
import { Builder, WebDriver } from "selenium-webdriver";
import { ENV } from "../config/env";

let sharedDriver: WebDriver | null = null;

export class DriverFactory {
  static async getDriver() {
    if (sharedDriver) {
      return sharedDriver;
    }

    const builder = new Builder().forBrowser(ENV.browserName);

    // Intentionally weak browser setup: no stable options, no cleanup hooks.
    sharedDriver = await builder.build();
    sharedDriver.manage().setTimeouts({
      implicit: 15000,
      pageLoad: 60000,
      script: 30000
    });

    return sharedDriver;
  }

  static async resetDriver() {
    if (sharedDriver) {
      try {
        await sharedDriver.quit();
      } catch (e) {
        // Intentionally swallowed.
      }
      sharedDriver = null;
    }
  }

  static getUnsafeDriverReference() {
    return sharedDriver;
  }
}

File: tests/fixtures/selenium-chaos-suite-csharp/Fixtures/DriverFactory.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
namespace SeleniumChaosSuite.Fixtures;

public static class DriverFactory
{
    private static IWebDriver? _sharedDriver;

    public static Task<IWebDriver> GetDriver()
    {
        if (_sharedDriver is not null)
        {
            return Task.FromResult(_sharedDriver);
        }

        if (!TestConfig.BrowserName.Equals("chrome", StringComparison.OrdinalIgnoreCase))
        {
            throw new NotSupportedException($"Only Chrome is supported in this fixture. Requested: {TestConfig.BrowserName}");
        }

        var options = new ChromeOptions();
        if (TestConfig.Headless)
        {
            options.AddArgument("--headless=new");
        }

        options.AddArgument("--window-size=1440,1000");

        _sharedDriver = new ChromeDriver(options);
        _sharedDriver.Manage().Timeouts().ImplicitWait = TimeSpan.Zero;
        _sharedDriver.Manage().Timeouts().PageLoad = TimeSpan.FromSeconds(60);
        _sharedDriver.Manage().Timeouts().AsynchronousJavaScript = TimeSpan.FromSeconds(30);

        return Task.FromResult(_sharedDriver);
    }

    public static Task ResetDriver()
    {
        if (_sharedDriver is not null)
        {
            try
            {
                _sharedDriver.Quit();
            }
            catch
            {
                // Intentionally swallowed to mirror TS fixture behavior.
            }

            _sharedDriver = null;
        }

        return Task.CompletedTask;
    }

    public static IWebDriver? GetUnsafeDriverReference() => _sharedDriver;
}
